### DISCUSSION TITLE
__fastcall__ functions keep storing parameters on the stack

### DISCUSSION THREAD

**Discussion by gorlik** [cc65/cc65#1590] [Q&A]
fastcall functions pass the first parameter using AX, however the prologue of the functions always store this value in the sw stack (with a matching incsp in the epilogue) regardless of the complexity of the function.

`void  __fastcall__ func4(unsigned char a)`
`{`
`  *((unsigned char *)0xd000)=a;`
`}`

gets compiled to:
`	jsr     pusha`
`	ldy     #$00`
`	lda     (sp),y`
`	sta     $D000`
`	jmp     incsp1`
instead of just:
`     sta $d000`
`     rts`

Can this be fixed with the peephole optimizer? Is adding a new attribute for the function a better way?
thanks

**Comment by mrdudz** [comment]
Did you compile with -O ?

**Comment by gorlik** [comment]
yes. Neither -O nor -Oisr will make the compiler generate different code.

**Comment by groessler** [comment]
I guess the compiler just saves the value of A on the stack, since it's quite volatile. Think about a function like

```
void __fastcall__ func4(unsigned char a)
{
... /* do many things */
*((unsigned char *)0xd000)=a;
}
```

The compiler doesn't know that in your case A is still valid. An improvement could be that it detects this case, but I don't think it's worth the effort. Such a simple function could as well be implemented in assembler, even by one which doesn't know much about assembler. Just a STA followed by a RTS.

**Reply by polluks** [reply]
That's exactly the job of the optimizer!
But we have not enough manpower for the optimizer in the cc65 team :-(

**Reply by gorlik** [reply]
What about adding a "naked" attribute? I hacked up some changes as a proof of concept.
This will allow to achieve best performance using inline assembly without resorting to assembly files.
Using inline in a C function is especially useful when accessing C defines which are not visible from the assembler.
If naked is an acceptable solution I am happy to cleanup the code and submit a pull request.

**Reply by MonteCarlos** [reply]
Is there some corresponding gcc attribute? We should not reinvent the wheel, if possible.

**Reply by groessler** [reply]
> without resorting to assembly files.

What's bad with that?

IMO inline assembly is fine if you want to add one or two instructions in a C function. But where is the benefit of using inline assembly for whole functions instead of using ca65?

**Reply by gorlik** [reply]
Sdcc, gcc, clang and msvc all have a naked attribute. 

The drawback of using separate assembly files is the difficulty of sharing constants and structures.
I.e. define a struct in the c file. Implement  most of the function that access the struct in C. Now add a couple of assembly functions in a separate file.
How do you keep them in sync? 

In my limited use of cc65 this happened multiple times.

**Reply by groessler** [reply]
> The drawback of using separate assembly files is the difficulty of sharing constants and structures.

Fair enough.

**Reply by mrdudz** [reply]
> What about adding a "naked" attribute? I hacked up some changes as a proof of concept. This will allow to achieve best performance using inline assembly without resorting to assembly files. Using inline in a C function is especially useful when accessing C defines which are not visible from the assembler. If naked is an acceptable solution I am happy to cleanup the code and submit a pull request.

What happened to that idea? I am totally willing to accept a PR that implements __naked__ in the same way __fastcall__ is implemented - please include testcases (i am especially curious about combining __naked__ and __fastcall__, for example)

**Comment by mrdudz** [comment]
> Is there some corresponding gcc attribute?

"naked" :)

it makes sense to me to have this
